July 16 – Temperature Analysis
=================
#Temperature Conversions
* Measuring Temperature
* Python as a calculator (Temperature Conversions)
* Standalone Python scripts (.py files)
* Control Structures: (for loops, ifs)
* Storing data (strings, lists, dictionaries, arrays)

#Time series data (Temperature for the past year)
* Using Python libraries
* Reading in temperature data from files ( pandas)
* Fancy indexing
* Aggregating data based on some condition

CWK:  Read in monthly 2015 precipitation values for New York, Philadelphia, Newark, and Boston data. Compare the temperature and precipitation values between cities. Submit a paragraph describing any patterns you find. You must support all conclusions with specific details (numbers, differences between two quantities, etc.)

Measuring Temperature
------------------------

1 Kelvin = 1 °Celsius = 9/5 °Fahrenheit

### From Kelvins:
* [°C] = [K] − 273.15
* [°F] = [K] × 9⁄5 − 459.67

### To Kelvins:
* [K] = [°C] + 273.15
* [K] = ([°F] + 459.67) × 5⁄9


In [1]:
#Let's convert 100 degrees Kelvin to Celsius:
100-273.15

-173.14999999999998

In [5]:
#Now write a function kelvin2celsius to convert from Kelvins to Celsius

def kelvin2celsius(K):
    """Converts Kelvin Temperature to Celsius Temperature
    
        Parameters
        ----------
        K: double
            Kelvin Temperature
        
        Returns
        -------
        :double
            Celsius Temperature
    """
    return K-273.15

In [7]:
#And test the functions:
print kelvin2celsius(100)
#Why do you think the numbers differ? Is the function implemented correctly?


-173.15


#### Implement functions to do the following conversions and test:
* celsius2kelvin
* celsius2Fahrenheit
* fahrenheit2celsius

Is one unit better than the other? Why?

In [11]:
#lets make a list of Kelvin temperatures
ktemps = range(0,100,10)
#and now, lets make a lookup table of our kelvin temperatures:
k2c = dict() #empty dictionary

for k in ktemps: #loop over temperatures
    k2c[k] = kelvin2celsius(k) #store result of function in dictionary

print k2c

{0: -273.15, 70: -203.14999999999998, 40: -233.14999999999998, 10: -263.15, 80: -193.14999999999998, 50: -223.14999999999998, 20: -253.14999999999998, 90: -183.14999999999998, 60: -213.14999999999998, 30: -243.14999999999998}


Can you explain what k2c is and what it stores? Why are the converted values so similar?
Now create look up tables for Celsius and Fahrenheit to each of the other units. You may use the same range or vary as desired. Is the same range appropriate? Did the range for Kelvins make sense?

2015 Temperature Data
------------------------
Lets explore temperature data for the past year from New York, Newark, Phildelphia, and Boston. The data comes from http://www.ncdc.noaa.gov/cdo-web/datasets and is stored in a CSV file that we will open using the pandas (http://pandas.pydata.org/) library. 

### Locations: 
* CITY:US360019 - New York, NY US
* CITY:US250002 - Boston, MA US
* CITY:US420015 - Philadelphia, PA US
* CITY:US340003 - Newark, NJ US

### Data Types: 	
* MNTM - Monthly mean temperature
* MMNT - Monthly Mean minimum temperature
* MMXT - Monthly Mean maximum temperature
* EMNT - Extreme minimum daily temperature
* EMXT - Extreme maximum daily temperature
* TPCP - Total precipitation
* EMXP - Extreme maximum daily precipitation
* TSNW - Total snow fall
* MXSD - Maximum snow depth

On blackboard there's more data documentation, so take a moment to review it

In [16]:
#lets open the file using pandas
import pandas as pd

#this will vary depending on where you put the data
#windows requires \\ for path seperation and to escape \
data_path = "C:\\Users\\story645\\Desktop\\EAS213_2015\\571549.csv"

df = pd.read_csv(data_path)

In [17]:
#lets see what's in dfile:
print df

                STATION                           STATION_NAME  ELEVATION  \
0     GHCND:US1NJBG0023                  OAKLAND 0.9 SSE NJ US      149.4   
1     GHCND:US1NJBG0023                  OAKLAND 0.9 SSE NJ US      149.4   
2     GHCND:US1NJBG0023                  OAKLAND 0.9 SSE NJ US      149.4   
3     GHCND:US1NJBG0023                  OAKLAND 0.9 SSE NJ US      149.4   
4     GHCND:US1NJBG0023                  OAKLAND 0.9 SSE NJ US      149.4   
5     GHCND:US1NJBG0023                  OAKLAND 0.9 SSE NJ US      149.4   
6     GHCND:US1PAMT0078                 BLUE BELL 0.8 NE PA US      106.7   
7     GHCND:US1PAMT0078                 BLUE BELL 0.8 NE PA US      106.7   
8     GHCND:US1PAMT0078                 BLUE BELL 0.8 NE PA US      106.7   
9     GHCND:US1PAMT0078                 BLUE BELL 0.8 NE PA US      106.7   
10    GHCND:US1PAMT0078                 BLUE BELL 0.8 NE PA US      106.7   
11    GHCND:US1PAMT0078                 BLUE BELL 0.8 NE PA US      106.7   

In [18]:
#that's kind of hard to see, so lets step back and see what fields are available:
print df.keys()

Index([u'STATION', u'STATION_NAME', u'ELEVATION', u'LATITUDE', u'LONGITUDE', u'DATE', u'EMXP', u'MXSD', u'TPCP', u'TSNW', u'EMXT', u'EMNT', u'MMXT', u'MMNT', u'MNTM'], dtype='object')


In [19]:
#How do we get just the New York temperatures? 
print df[u'STATION_NAME']

0           OAKLAND 0.9 SSE NJ US
1           OAKLAND 0.9 SSE NJ US
2           OAKLAND 0.9 SSE NJ US
3           OAKLAND 0.9 SSE NJ US
4           OAKLAND 0.9 SSE NJ US
5           OAKLAND 0.9 SSE NJ US
6          BLUE BELL 0.8 NE PA US
7          BLUE BELL 0.8 NE PA US
8          BLUE BELL 0.8 NE PA US
9          BLUE BELL 0.8 NE PA US
10         BLUE BELL 0.8 NE PA US
11         BLUE BELL 0.8 NE PA US
12    PERKIOMENVILLE 4.6 SW PA US
13    PERKIOMENVILLE 4.6 SW PA US
14    PERKIOMENVILLE 4.6 SW PA US
...
1213                   EDISON TWP 1.9 N NJ US
1214                   EDISON TWP 1.9 N NJ US
1215                   EDISON TWP 1.9 N NJ US
1216    PHILADELPHIA FRANKLIN INSTITUTE PA US
1217    PHILADELPHIA FRANKLIN INSTITUTE PA US
1218    PHILADELPHIA FRANKLIN INSTITUTE PA US
1219    PHILADELPHIA FRANKLIN INSTITUTE PA US
1220    PHILADELPHIA FRANKLIN INSTITUTE PA US
1221    PHILADELPHIA FRANKLIN INSTITUTE PA US
1222                           BROCKTON MA US
1223                      

In [20]:
#So what do NYC stations have in common? They're in NY!
#So lets figure out which station names are in Y:
#Python allows for partial checks:
print 'NY' in 'BROOKLYN 3.1 NW NY US'

True


In [21]:
#Python list comprehension
for food in ['apples', 'grapes', 'oranges']:
    print food

apples
grapes
oranges


In [22]:
#So lets extend to all the stations in the list
for sn in df['STATION_NAME']:
    print sn, 'NY' in sn

OAKLAND 0.9 SSE NJ US False
OAKLAND 0.9 SSE NJ US False
OAKLAND 0.9 SSE NJ US False
OAKLAND 0.9 SSE NJ US False
OAKLAND 0.9 SSE NJ US False
OAKLAND 0.9 SSE NJ US False
BLUE BELL 0.8 NE PA US False
BLUE BELL 0.8 NE PA US False
BLUE BELL 0.8 NE PA US False
BLUE BELL 0.8 NE PA US False
BLUE BELL 0.8 NE PA US False
BLUE BELL 0.8 NE PA US False
PERKIOMENVILLE 4.6 SW PA US False
PERKIOMENVILLE 4.6 SW PA US False
PERKIOMENVILLE 4.6 SW PA US False
PERKIOMENVILLE 4.6 SW PA US False
PERKIOMENVILLE 4.6 SW PA US False
PERKIOMENVILLE 4.6 SW PA US False
EDISON TWP 1.4 N NJ US False
EDISON TWP 1.4 N NJ US False
EDISON TWP 1.4 N NJ US False
SAYREVILLE 1.3 ESE NJ US False
SAYREVILLE 1.3 ESE NJ US False
SAYREVILLE 1.3 ESE NJ US False
SELLERSVILLE PA US False
SELLERSVILLE PA US False
SELLERSVILLE PA US False
SELLERSVILLE PA US False
SELLERSVILLE PA US False
SELLERSVILLE PA US False
MAPLEWOOD TWP 0.9 SE NJ US False
MAPLEWOOD TWP 0.9 SE NJ US False
MAPLEWOOD TWP 0.9 SE NJ US False
MAPLEWOOD TWP 0.9 SE NJ U

In [23]:
#So lets extend to all the stations in the list
nyc_inds = ['NY' in sn for sn in df['STATION_NAME']]

print len(nyc_inds), len(df)

1228 1228


Lets deconstruct:
* sn in df['STATION_NAME'] is a list comprehension (for loop)
* so sn is sequentially each station in the list of names (1228 in all)
* 'NY' in sn is a boolean equation
* returns true when the 'NY' string is in the the station name
* nyc_inds is a list of boolean values: true when the station is in NY and false otherwise

In [24]:
import numpy as np
inds = np.array([True, True, False, False, True])
a = np.arange(5)
print "A:", a
print a[inds]
inds2 = np.array([False, False, False, False, True])
print a[inds2]
inds3 = np.array([True, True, True, True, False])
print a[inds3]

A: [0 1 2 3 4]
[0 1 4]
[4]
[0 1 2 3]


In [25]:
#so how can it be used? Fancye indexing! 
# http://pandas.pydata.org/pandas-docs/stable/indexing.html

nyc = df[nyc_inds] # this grabs only the rows for which nyc_inds is true

In [26]:
print nyc

                STATION                                      STATION_NAME  \
101   GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
102   GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
103   GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
104   GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
105   GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
106   GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
170   GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
171   GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
172   GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
173   GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
174   GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
175   GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   

In [27]:
#oops, how did  NESHAMINY FALLS PA US get in there?
print "NY" in  "NESHAMINY FALLS PA US"

True


In [28]:
#SO how to correct:
print "NY US" in "NESHAMINY FALLS PA US"
print "NY US" in "BROOKLYN 3.1 NW NY US"
#and to be even safer:
print "BROOKLYN 3.1 NW NY US".endswith("NY US")

nyc_inds = [sn.endswith("NY US") for sn in df['STATION_NAME']]
nyc = df[nyc_inds]
print nyc

False
True
True
               STATION                                      STATION_NAME  \
101  GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
102  GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
103  GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
104  GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
105  GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
106  GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
170  GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
171  GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
172  GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
173  GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
174  GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
175  GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US

In [29]:
#so how do we start grouping things? What variable makes sense? Why?
#pandas makes this easy with grouby: 
# http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html
nyc_groups = nyc.groupby('DATE')
print nyc_groups

In [30]:
#Huh, so how to look throught that? The easiest is to make it a list
print list(nyc_groups)

[(20150101,                STATION                                      STATION_NAME  \
101  GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
170  GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
234  GHCND:USW00014732                  NEW YORK LAGUARDIA AIRPORT NY US   
409  GHCND:USW00054787                FARMINGDALE REPUBLIC AIRPORT NY US   
501  GHCND:US1NYRL0005                          WEST NYACK 1.3 WSW NY US   
603  GHCND:USW00094728   NEW YORK CENTRAL PARK OBS BELVEDERE TOWER NY US   
627  GHCND:USC00301309                                  CENTERPORT NY US   
639  GHCND:US1NYQN0002                       MIDDLE VILLAGE 0.5 SW NY US   
645  GHCND:US1NYNS0007                           FLORAL PARK 0.4 W NY US   
687  GHCND:USW00094745                      WESTCHESTER CO AIRPORT NY US   
734  GHCND:US1NYRC0002                        STATEN ISLAND 1.4 SE NY US   
740  GHCND:US1NYRC0001                       STATEN ISLAND 4.5 SSE NY US   


In [31]:
#but how does this help? Well we want one value for each day, right? So...

print list(nyc.groupby('DATE'))

[(20150101,                STATION                                      STATION_NAME  \
101  GHCND:USW00094789  NEW YORK J F KENNEDY INTERNATIONAL AIRPORT NY US   
170  GHCND:US1NYWC0003                        WHITE PLAINS 3.1 NNW NY US   
234  GHCND:USW00014732                  NEW YORK LAGUARDIA AIRPORT NY US   
409  GHCND:USW00054787                FARMINGDALE REPUBLIC AIRPORT NY US   
501  GHCND:US1NYRL0005                          WEST NYACK 1.3 WSW NY US   
603  GHCND:USW00094728   NEW YORK CENTRAL PARK OBS BELVEDERE TOWER NY US   
627  GHCND:USC00301309                                  CENTERPORT NY US   
639  GHCND:US1NYQN0002                       MIDDLE VILLAGE 0.5 SW NY US   
645  GHCND:US1NYNS0007                           FLORAL PARK 0.4 W NY US   
687  GHCND:USW00094745                      WESTCHESTER CO AIRPORT NY US   
734  GHCND:US1NYRC0002                        STATEN ISLAND 1.4 SE NY US   
740  GHCND:US1NYRC0001                       STATEN ISLAND 4.5 SSE NY US   


In [32]:
nyc_groups = nyc.groupby('DATE').mean()
print nyc_groups

          ELEVATION   LATITUDE  LONGITUDE        EMXP         MXSD  \
DATE                                                                 
20150101  34.069231  40.789296 -73.825647  461.000000 -2122.615385   
20150201  34.069231  40.789296 -73.825647  263.538462 -5266.076923   
20150301  34.069231  40.789296 -73.825647  248.000000 -3694.384615   
20150401  34.069231  40.789296 -73.825647  289.769231 -6153.230769   
20150501  34.069231  40.789296 -73.825647  165.923077 -6922.384615   
20150601  33.571429  40.784668 -73.803001  366.500000 -7142.142857   

                 TPCP         TSNW         EMXT         EMNT         MMXT  \
DATE                                                                        
20150101  1258.230769 -1241.230769 -5317.384615 -5447.692308 -5374.538462   
20150201   497.692308 -1354.615385 -5358.846154 -5463.692308 -5387.153846   
20150301  1159.000000 -1270.076923 -5313.923077 -5438.846154 -5353.307692   
20150401   487.307692 -1538.307692 -5265.692308 -5386.

In [31]:
nyc[nyc==-9999]=np.nan #spreadsheet has words and numbers

TypeError: Cannot do boolean setting on mixed-type frame

In [34]:
#Those numbers are a bit high!
import numpy as np

datakeys = [u'EMXP', u'MXSD', u'TPCP', u'TSNW', u'EMXT', u'EMNT', u'MMXT', u'MMNT', u'MNTM']

In [37]:
for dk in datakeys:
    print dk, nyc[dk]

EMXP 101    411
102    300
103    282
104    221
105    104
106    277
170    493
171     74
172    231
173    587
174    145
175    132
234    503
235    274
236    259
...
737    262
738     97
739    582
740    282
741    472
742    191
743    193
744     56
745    488
838    490
839     60
840    248
841     33
842      8
843     89
Name: EMXP, Length: 79, dtype: int64
MXSD 101     250
102     200
103     180
104       0
105       0
106       0
170   -9999
171   -9999
172   -9999
173   -9999
174   -9999
175   -9999
234     230
235     230
236     280
...
737   -9999
738   -9999
739   -9999
740     249
741     127
742     241
743   -9999
744   -9999
745   -9999
838      25
839     241
840     279
841   -9999
842   -9999
843   -9999
Name: MXSD, Length: 79, dtype: int64
TPCP 101    1350
102     599
103    1239
104     409
105     117
106    1267
170     509
171      77
172     856
173     775
174     162
175     311
234    1265
235     534
236    1155
...
737     371
738     169
739  

In [38]:
for dk in datakeys:
    print dk, nyc[dk]==-9999

EMXP 101    False
102    False
103    False
104    False
105    False
106    False
170    False
171    False
172    False
173    False
174    False
175    False
234    False
235    False
236    False
...
737    False
738    False
739    False
740    False
741    False
742    False
743    False
744    False
745    False
838    False
839    False
840    False
841    False
842    False
843    False
Name: EMXP, Length: 79, dtype: bool
MXSD 101    False
102    False
103    False
104    False
105    False
106    False
170     True
171     True
172     True
173     True
174     True
175     True
234    False
235    False
236    False
...
737     True
738     True
739     True
740    False
741    False
742    False
743     True
744     True
745     True
838    False
839    False
840    False
841     True
842     True
843     True
Name: MXSD, Length: 79, dtype: bool
TPCP 101    False
102    False
103    False
104    False
105    False
106    False
170    False
171    False
172    False
173    F

In [39]:
for dk in datakeys:
    print dk, nyc[dk][nyc[dk]==-9999]

EMXP Series([], name: EMXP, dtype: int64)
MXSD 170   -9999
171   -9999
172   -9999
173   -9999
174   -9999
175   -9999
409   -9999
410   -9999
411   -9999
412   -9999
413   -9999
414   -9999
502   -9999
503   -9999
504   -9999
505   -9999
506   -9999
633   -9999
640   -9999
641   -9999
642   -9999
643   -9999
644   -9999
646   -9999
649   -9999
650   -9999
687   -9999
688   -9999
689   -9999
690   -9999
691   -9999
692   -9999
735   -9999
737   -9999
738   -9999
739   -9999
743   -9999
744   -9999
745   -9999
841   -9999
842   -9999
843   -9999
Name: MXSD, dtype: int64
TPCP Series([], name: TPCP, dtype: int64)
TSNW 409   -9999
410   -9999
411   -9999
412   -9999
413   -9999
414   -9999
633   -9999
687   -9999
688   -9999
689   -9999
690   -9999
691   -9999
692   -9999
Name: TSNW, dtype: int64
EMXT 170   -9999
171   -9999
172   -9999
173   -9999
174   -9999
175   -9999
501   -9999
502   -9999
503   -9999
504   -9999
505   -9999
506   -9999
633   -9999
639   -9999
640   -9999
641   -9999

In [40]:
for dk in datakeys:
    nyc[dk][nyc[dk]==-9999] = np.nan #assigns not a number

C:\Users\story645\Anaconda\lib\site-packages\pandas\core\series.py:632: SettingWithCopyWarning: A value is trying to be set on a copy of a slice from a DataFrame
  self.where(~key, value, inplace=True)


In [41]:
nyc_groups = nyc.groupby('DATE').mean()
print nyc_groups
nyc_groups.to_csv("new_york.csv") #save data out to csv

          ELEVATION   LATITUDE  LONGITUDE        EMXP        MXSD  \
DATE                                                                
20150101  34.069231  40.789296 -73.825647  461.000000  240.300000   
20150201  34.069231  40.789296 -73.825647  263.538462  255.666667   
20150301  34.069231  40.789296 -73.825647  248.000000  246.000000   
20150401  34.069231  40.789296 -73.825647  289.769231    0.000000   
20150501  34.069231  40.789296 -73.825647  165.923077    0.000000   
20150601  33.571429  40.784668 -73.803001  366.500000    0.000000   

                 TPCP        TSNW        EMXT        EMNT        MMXT  \
DATE                                                                    
20150101  1258.230769  351.090909  144.500000 -137.833333   20.666667   
20150201   497.692308  217.090909   54.666667 -172.500000   -6.666667   
20150301  1159.000000  317.000000  152.000000 -118.666667   66.666667   
20150401   487.307692    0.000000  256.500000   -4.333333  158.000000   
20150501 

Now repeat for Boston, Philadelphia, and Newark

Then:
Read in monthly 2015 precipitation values for New York, Philadelphia, Newark, and Boston data. Compare the temperature and precipitation values between cities. Submit a paragraph describing any patterns you find. You must support all conclusions with specific details (numbers, differences between two quantities, etc.)